autopush/router/apns2.py

```diff
@@ -166,19 +178,52 @@ def send(self, router_token, payload, apns_id,
             finally:
                 # Returning a closed connection to the pool is ok.
                 # hyper will reconnect on .request()
-                self._return_connection(connection)
+                if connection:
+                    self._return_connection(connection)
 
     def _get_connection(self):
-        try:
-            connection = self.connections.pop()
-            return connection
-        except IndexError:
+        self._in_use_connections += 1
+        if self.log:
+            self.log.debug("Got New APNS connection")
+        if self._in_use_connections > self._max_connections:
             raise RouterException(
```

Comment:
This exception should happen before we increment `_in_use_connections`, because it currently will leave that value in an invalid state. Probably should gate the exception on `if self._in_use_connections == self._max_connections`.